clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -537,51 +586,137 @@ private void updateGroupSubscription(Set<String> topics) {
+
+            throw new IllegalStateException("Assignor supporting the COOPERATIVE protocol violates its requirements");
+        }
+    }
+
     @Override
     protected void onJoinPrepare(int generation, String memberId) {
         // commit offsets prior to rebalance if auto-commit enabled
         maybeAutoCommitOffsetsSync(time.timer(rebalanceConfig.rebalanceTimeoutMs));
 
         // execute the user's callback before rebalance
-        ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
-
+        final Set<TopicPartition> revokedPartitions;
+        final AtomicReference<Exception> firstException = new AtomicReference<>(null);
         switch (protocol) {
             case EAGER:
-                Set<TopicPartition> revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
-                log.info("Revoking previously assigned partitions {}", revokedPartitions);
-                try {
-                    listener.onPartitionsRevoked(revokedPartitions);
-                } catch (WakeupException | InterruptException e) {
-                    throw e;
-                } catch (Exception e) {
-                    log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
-                }
-
-                // also clear the assigned partitions since all have been revoked
+                // revoke all partitions
+                revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
                 subscriptions.assignFromSubscribed(Collections.emptySet());
 
+                firstException.compareAndSet(null, maybeInvokePartitionsRevoked(revokedPartitions));
                 break;
 
             case COOPERATIVE:
+                // only revoke those partitions that are not in the subscription any more.
+                Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                revokedPartitions = ownedPartitions.stream()
+                    .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
+                    .collect(Collectors.toSet());
+
+                if (!revokedPartitions.isEmpty()) {
+                    ownedPartitions.removeAll(revokedPartitions);
+                    subscriptions.assignFromSubscribed(ownedPartitions);
+
+                    firstException.compareAndSet(null, maybeInvokePartitionsRevoked(revokedPartitions));
+                }
+
                 break;
         }
 
         isLeader = false;
         subscriptions.resetGroupSubscription();
+
+        if (firstException.get() != null) {
+            throw new KafkaException("User rebalance callback throws an error", firstException.get());
+        }
     }
 
+    /**
+     * @throws KafkaException if the rebalance callback throws exception
+     */
+    @Override
+    public void resetGeneration(String errorMessage) {
```

Comment:
Hmm.. This method is invoked from `unsubscribe`. It seems like we might want to use `onPartitionsRevoked` in that case because it can still be a graceful handoff.

Reply:
This is also be invoked when encountering errors from heartbeat / join-group etc, in which case we should really invoke `onPartitionsLost`.

I think I'll have to duplicate the logic a bit to distinguish these two cases.